src/backend/access/transam/xlog.c

```diff
@@ -8814,11 +8828,31 @@ GetInsertRecPtr(void)
  * GetLastWrittenPageLSN -- Returns maximal LSN of written page
  */
 XLogRecPtr
-GetLastWrittenPageLSN(void)
+GetLastWrittenPageLSN(RelFileNode *rnode)
 {
 	XLogRecPtr lsn;
 	SpinLockAcquire(&XLogCtl->info_lck);
 	lsn = XLogCtl->lastWrittenPageLSN;
+	if (rnode != NULL)
+	{
+		for (int i = 0; i < LAST_WRITTEN_CACHE_SIZE; i++)
+		{
+			if (RelFileNodeEquals(*rnode, XLogCtl->lastWrittenPageCache[i].rnode))
+			{
+				lsn = XLogCtl->lastWrittenPageCache[i].lsn;
+				break;
+			}
+		}
+	}
+	else
+	{
+		/* Find maximum of all cached LSNs */
+		for (int i = 0; i < LAST_WRITTEN_CACHE_SIZE; i++)
+		{
+			if (XLogCtl->lastWrittenPageCache[i].lsn > lsn)
+				lsn = XLogCtl->lastWrittenPageCache[i].lsn;
```

Comment:
Why not max(XLogCtl->lastWrittenPageLSN, max_cached) ?
This logic deserves some comment.